<a href="https://colab.research.google.com/github/baldalerie/Home-work/blob/main/%D0%9B%D1%8B%D1%81%D0%B8%D0%BA%D0%BE%D0%B2_%D0%92_%D0%A1_%D0%BA%D1%802.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Задача 1

In [1]:
import scipy.stats as stats
import pandas as pd
from random import randint

alpha = 0.01 # уровень значимости
X = 22 # необходимая сумма очков
s_N = []
for i in range(1,100000): # расширяем выборку до 100000 подбрасываний дайсов
    d1 = randint(1,6) # равномерное распределение p=1/6
    d2 = randint(1,6)
    d3 = randint(1,6)
    d4 = randint(1,6)
    s = d1 + d2 + d3 + d4
    s_N.append(s)

In [2]:
s_N = pd.Series(s_N) # упорядочиваем список сумм
stats.zscore(s_N) # стандартизируем список по нормальному распределению (с правилом трех сигм)
print(s_N)

0        11
1         8
2        14
3        19
4        13
         ..
99994    17
99995    14
99996    11
99997    18
99998    19
Length: 99999, dtype: int64


In [3]:
# Вычисляем значение z - на сколько стандартных отклонений случайная величина отличается от мат. ожидания
z = (X - s_N.mean()) / s_N.std()
print('z = ', round(z, 4))

z =  2.3398


In [4]:
# В соответсвии со значением z находим p
p_value = stats.norm.sf(abs(z))
# Вероятность случайной величины (сумма очков) реализоваться выше, чем z
print('p_value = ', round(p_value, 4))

p_value =  0.0096


In [5]:
if p_value < alpha:
    print('Отклонить гипотезу H0')
else:
    print('Не отклонить гипотезу H0')

Отклонить гипотезу H0


Вывод: отклоним гипотезу H0 в пользу альтернативной Н1. Игрок пройдет проверку (сумма очков на дайсах равна 22 и больше) при гипотезе Н1

Задача 2

In [7]:
import scipy.stats as stats
from mpmath import *
mp.dps = 100000 # устанавливаем количество знаков после запятой

#alpha = 0.9 - уровень доверия
N = 100000

pi = 4*atan(1)
#print(pi)

In [10]:
data = pi
data = str(data)

In [11]:
# Найдем эмпирическую статистику цифр
exp_list = []
for i in range(0,10):
    i = str(i)
    s = data.count(i)
    exp_list.append(s)
print('Экспериментальная частота цифр 1-9', 'составляет', exp_list, 'соответственно.')

Экспериментальная частота цифр 1-9 составляет [9999, 10137, 9908, 10026, 9970, 10027, 10028, 10025, 9978, 9902] соответственно.


In [8]:
# Для нулевой гипотезы распределение цифр в числе пи равномерное, т.е вероятность вхождения каждой цифры 1/10
teor_list = []
p = 0.1
for i in range(0,10):
    i = str(i)
    s = N * p
    s = int(s)
    teor_list.append(s)
print('Теоретическая частота цифр 1-9', 'составляет', teor_list, 'соответственно.')

Теоретическая частота цифр 1-9 составляет [10000, 10000, 10000, 10000, 10000, 10000, 10000, 10000, 10000, 10000] соответственно.


In [12]:
# Используем критерий хи-квадрат для вычисления максимального уровня значимости
stat, p = stats.chisquare(exp_list, teor_list)
print('Значение статистики = %.3f, p-value = %.3f' % (stat, p.round(3)))

Значение статистики = 4.104, p-value = 0.904


Вывод: максимальный уровень значимости, на котором принимается нулевая гипотеза равен 0.904